File: abc244/f/bfs.cc
// https://atcoder.jp/contests/abc244/tasks/abc244_f

#include <climits>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int main() {
  const int INF = INT_MAX;
  int N, M;
  cin >> N >> M;
  int N2 = 1<<N;
  vector<vector<int> > graph(N);
  for (int m = 0; m < M; m++) {
    int U, V;
    cin >> U >> V;
    U--;
    V--;
    graph.at(U).push_back(V);
    graph.at(V).push_back(U);
  }
  queue<pair<int, int> > todo;
  vector<vector<int> > dist(N2, vector<int>(N, INF));
  todo.push(make_pair(0, 0));
  dist[0][0] = 0;
  for (int v = 0; v < N; v++) {
    todo.push(make_pair(1<<v, v));
    dist[1<<v][v] = 1;
  }
  while (!todo.empty()) {
    int b = todo.front().first;
    int v = todo.front().second;
    todo.pop();
    for (size_t i = 0; i < graph[v].size(); i++) {
      int nv = graph[v][i];
      int nb = b ^ (1<<nv);
      if (dist[nb][nv] == INF) {
        dist[nb][nv] = dist[b][v] + 1;
        todo.push(make_pair(nb, nv));
      }
    }
  }
  long long ans = 0;
  for (int b = 0; b < N2; b++) {
    int now = INF;
    for (int v = 0; v < N; v++) {
      now = min(now, dist[b][v]);
    }
    ans += now;
  }
  cout << ans << endl;
  return 0;
}

File: abc257/f/bfs.py
# https://atcoder.jp/contests/abc257/tasks/abc257_f

import queue

N, M = map(int, input().split())
graph = [[] for n in range(N + 1)]  # 1-indexed
for m in range(M):
    U, V = map(int, input().split())
    graph[U].append(V)
    graph[V].append(U)

A = {}
INF = 3 * 10**5 + 2
A[1] = [INF] * (N + 1)  # 1からの最短距離
A[N] = [INF] * (N + 1)  # Nからの最短距離
for start in (1, N):
    dist = A[start]
    dist[start] = 0
    todo = queue.Queue()
    todo.put(start)
    while not todo.empty():
        u = todo.get()
        for v in graph[u]:
            if dist[v] == INF:
                dist[v] = dist[u] + 1
                todo.put(v)
ans = []
for i in range(1, N + 1):
    now = min(A[1][N], A[1][0] + A[N][i], A[1][i] + A[N][0])
    ans.append(-1 if now >= INF else now)
print(' '.join(map(str, ans)))
